go/server/handlers.go

```diff
@@ -61,9 +62,104 @@ func (s *Server) homeHandler(c *gin.Context) {
 	})
 }
 
-func (s *Server) searchCompareHandler(c *gin.Context) {
-	c.HTML(http.StatusOK, "search_compare.tmpl", gin.H{
-		"title": "Vitess benchmark",
+func (s *Server) compareHandler(c *gin.Context) {
+	reference := c.Query("r")
+	compare := c.Query("c")
+
+	compareSHA := map[string]interface{}{
+		"SHA":   compare,
+		"short": git.ShortenSHA(compare),
+	}
+	referenceSHA := map[string]interface{}{
+		"SHA":   reference,
+		"short": git.ShortenSHA(reference),
+	}
+	if reference == "" || compare == "" {
+		c.HTML(http.StatusOK, "compare.tmpl", gin.H{
+			"title":     "Vitess benchmark",
+			"reference": referenceSHA,
+			"compare":   compareSHA,
+		})
+		return
+	}
+
+	SHAs := []string{reference, compare}
+
+	// Get macro benchmarks from all the different types
+	macros := map[string]map[macrobench.Type]macrobench.DetailsArray{}
+	for _, sha := range SHAs {
+		macros[sha] = map[macrobench.Type]macrobench.DetailsArray{}
+		for _, mtype := range macrobench.Types {
+			macro, err := macrobench.GetResultsForGitRef(mtype, sha, s.dbClient)
+			if err != nil {
+				handleRenderErrors(c, err)
+				return
+			}
+			macros[sha][mtype] = macro.ReduceSimpleMedian()
+		}
+	}
+	macrosMatrixes := map[macrobench.Type]interface{}{}
+	for _, mtype := range macrobench.Types {
+		macrosMatrixes[mtype] = macrobench.CompareDetailsArrays(macros[reference][mtype], macros[compare][mtype])
+	}
+
+	// compare micro benchmarks
+	micros := map[string]microbench.MicroBenchmarkDetailsArray{}
+	for _, sha := range SHAs {
+		micro, err := microbench.GetResultsForGitRef(sha, s.dbClient)
+		if err != nil {
+			handleRenderErrors(c, err)
+			return
+		}
+		micros[sha] = micro.ReduceSimpleMedian()
+	}
+	microsMatrix := microbench.MergeMicroBenchmarkDetails(micros[reference], micros[compare])
+	sort.SliceStable(microsMatrix, func(i, j int) bool {
+		return !(microsMatrix[i].Current.NSPerOp < microsMatrix[j].Current.NSPerOp)
+	})
+
+	c.HTML(http.StatusOK, "compare.tmpl", gin.H{
+		"title":          "Vitess benchmark",
+		"reference":      referenceSHA,
+		"compare":        compareSHA,
+		"microbenchmark": microsMatrix,
+		"macrobenchmark": macrosMatrixes,
+	})
+}
+
+func (s *Server) searchHandler(c *gin.Context) {
+	search := c.Query("s")
+	if search == "" {
+		c.HTML(http.StatusOK, "search.tmpl", gin.H{
+			"title": "Vitess benchmark",
+		})
+		return
+	}
+
+	// Get macro benchmarks from all the different types
+	macros := map[macrobench.Type]interface{}{}
+	for _, mtype := range macrobench.Types {
```

Comment:
this is almost the same for loop that we have on line 92. wdyt about extracting a method that can be used from both places?

Reply:
Good idea, I have extracted the method via 20c6de4